# v3 — the patch

v1 and v2 were faithful reproductions with their defects corrected. This is the
step where the design is allowed to *change*. Four pieces:

| | |
|---|---|
| **A1** | the player metric — the per-pitch score becomes a swappable component and the choice is made on evidence |
| **A2** | pitch-frame features: batter-frame location, handedness, velocity and movement |
| **A3** | the take model — verify what it is actually learning |
| **A4** | personalization: replace the binary hot-zone flag with a continuous surface |

**How models are chosen.** Every model uses the shared early-stopping learner.
Between two variants, the selection rule decides:

1. **Accuracy** — paired held-out MSE on the same pitches, with a 95% interval
   from resampling games. A variant wins if it is better on swings without
   being worse on takes; an interval spanning 0 is a tie, and the simpler model
   stays.
2. **Calibration** checks the winner; if its slope interval excludes 1 it is
   recalibrated, never rejected.
3. **Guardrails** — the player-metric checks — are reported with intervals and
   can flag a change, but never choose one.

The metric choice (A1) is the exception: there the model is held fixed, so
accuracy cannot differ and the player checks are the only evidence.

Every figure is held out. Generic comparisons use the generic folds (training
from 2021); whenever a personalized variant is involved, every row uses the
personalized folds (training from 2022, 2021 as prior data only). 2026 is not
loaded.

In [1]:
import sys; sys.path.insert(0, '..')
import numpy as np, pandas as pd
import matplotlib.pyplot as plt, seaborn as sns

from src import data as D, evaluate as E, features as F, decision as DEC
from src.baselines import fit_direct, predict_both, train_early_stopped, V1_FEATURES, _dmatrix

pd.set_option('display.width', 170); pd.set_option('display.max_columns', 30)
sns.set_theme(style='whitegrid')
PALETTE = ['#FFB000','#648FFF','#785EF0','#DC267F','#FE6100','#3D1EB2']
SEASONS = range(2021, 2026)          # 2026 is kept out of model selection; not loaded here
SHOW, MODEL = E.SHOW, E.MODEL_COLS
ACC_COLS = ['pitches', 'diff_%', 'diff_%_lo', 'diff_%_hi', 'verdict']
PF = E.PERSONALIZED_FOLDS

# the full pitch frame, used from A2 onward
V3_FEATURES = ['plate_x_bat','plate_z_norm','count','stand','p_throws',
               'release_speed','pfx_x','pfx_z','pitch_type']

def pooled(a, b, labels):
    """Paired accuracy of run a against run b, pooled over the held-out seasons."""
    return E.paired_accuracy(a, b, labels).xs('pooled', level='season')[ACC_COLS]

In [2]:
df = D.drop_pitchers_batting(D.load_seasons(SEASONS, verbose=False))
df = D.add_zone_frame(df.dropna(subset=['plate_x_mid','plate_z_mid']))
for c in ('stand','p_throws','pitch_type'):
    df[c] = df[c].astype('category')
print(f'{len(df):,} pitches')

3,517,807 pitches


## A1 — which per-pitch score?

Every candidate is built from the same counterfactual pair, `q_swing` and
`q_take`, so the models are identical and only the aggregation differs. The
model here is v1's features on the generic folds, held fixed.

- `chosen_value` — the value of the action taken
- `signed_edge` — how much better the chosen action was than the alternative
- `regret` — `max(0, −signed_edge)`; zero whenever the hitter was right
- `close_weighted` — signed correctness, weighted by how close the call was
- `correct_decision` — +1 right, −1 wrong, no magnitude at all

**The selection criterion for a score, fixed before the numbers:** construct
validity first (does it punish chasing *and* reward attacking strikes, each
partialled on the other); reliability second as a floor; Zone% as a veto above
|r| ≈ 0.3; predictive validity reported but not decisive.

In [3]:
run_a1 = E.run_folds(df, V1_FEATURES, E.GENERIC_FOLDS, fit=fit_direct)
a1 = pd.DataFrame([{'variant': name, **E.metric_checks(run_a1.held, name)['summary']}
                   for name in DEC.SCORES]).set_index('variant')
a1[SHOW]

,chase | zone_swing,zone_swing | chase,split-half r,YoY R2 (mean),Zone% |r|,next-season r (partial)
variant,,,,,,
chosen_value,-0.889,0.681,0.765,0.537,0.124,0.126
signed_edge,-0.920,0.700,0.819,0.596,0.299,0.085
regret,-0.831,0.530,0.795,0.581,0.504,0.028
close_weighted,-0.701,0.802,0.606,0.317,0.045,0.101
correct_decision,-0.953,0.906,0.826,0.586,0.250,0.115


In [4]:
# close_weighted's kernel scale controls how narrowly it focuses on close calls.
h = run_a1.held
rows = []
for scale in [0.02, 0.05, 0.10, 0.20, 0.40, 0.80]:
    col = f'cw_{scale}'
    h[col] = DEC.close_weighted(h, scale=scale)
    rows.append({'scale': scale, **E.metric_checks(h, col)['summary']})
sweep = pd.DataFrame(rows).set_index('scale')
sc_cw, sc_cd = E.player_metric(h, 'cw_0.8'), E.player_metric(h, 'correct_decision')
m = sc_cw.merge(sc_cd, on=['season', 'batter'])
print(f'corr(close_weighted at 0.8, correct_decision) across hitter-seasons: '
      f'{np.corrcoef(m.raw_x, m.raw_y)[0, 1]:.3f}')
sweep[SHOW]

corr(close_weighted at 0.8, correct_decision) across hitter-seasons: 0.997


,chase | zone_swing,zone_swing | chase,split-half r,YoY R2 (mean),Zone% |r|,next-season r (partial)
scale,,,,,,
0.02,-0.290,0.581,0.284,0.105,0.104,0.063
0.05,-0.701,0.802,0.606,0.317,0.045,0.101
0.10,-0.853,0.859,0.717,0.433,0.123,0.116
0.20,-0.913,0.888,0.774,0.507,0.187,0.120
0.40,-0.937,0.900,0.802,0.548,0.220,0.119
0.80,-0.946,0.905,0.815,0.568,0.236,0.117


In [5]:
# What each score pays out, by how obvious the decision is (runs per pitch).
def region(g):
    inz = D.in_rulebook_zone(g, ball_edge=True)
    near = (g.plate_x_bat.abs() <= 1.15) & g.plate_z_norm.between(-0.35, 1.35)
    return pd.Series(np.where(inz, 'zone', np.where(near, 'shadow', 'chase')), index=g.index)

h['region'] = region(h)
h.groupby('region', observed=True)[['signed_edge', 'regret']].mean().round(4)

,signed_edge,regret
region,,
chase,0.0867,-0.0270
shadow,0.0110,-0.0334
zone,0.0506,-0.0164


**By the criterion fixed in advance, `correct_decision` wins.** It has the best
construct validity by a wide margin (−0.953 / +0.906, against −0.920 / +0.700
for `signed_edge`), the best split-half reliability, and passes the Zone% veto
(0.250) that `signed_edge` sits right on (0.299) and `regret` fails outright
(0.504). As the closeness kernel widens, `close_weighted` converges on
`correct_decision` (correlation 0.997 at scale 0.8), and construct validity
improves the whole way.

**The run-value magnitude is what carries the pitch-mix bias.** `signed_edge`
pays about eight times more per pitch for an obvious take (chase region, +0.087
runs) than for a genuinely close call (shadow, +0.011), so a hitter thrown more
junk scores higher for the same judgement.

**`signed_edge` stays selected, as a trade.** A sign-based score treats a
razor-thin call and an obvious blunder alike, and it barely registers a feature
that changes how much a swing is worth without flipping the decision. Every
published metric keeps the magnitude. A4 and the ladder below measure what each
choice gives up.

## A2 — pitch-frame features

Added in three groups, on the generic folds. Location moves to the batter frame
(positive = inside for every hitter) and height is normalised against the
common zone from listed height rather than the per-pitch operator bounds, which
carry 0.073–0.098 ft of measurement noise and change definition in 2026.

In [6]:
LADDER_A2 = {
  'v1 features':            V1_FEATURES,
  '+batter frame':          ['plate_x_bat','plate_z_norm','count'],
  '+handedness':            ['plate_x_bat','plate_z_norm','count','stand','p_throws'],
  '+pitch characteristics': V3_FEATURES,
}
a2_runs = {label: E.run_folds(df, feats, E.GENERIC_FOLDS, fit=fit_direct) for label, feats in LADDER_A2.items()}
a2 = pd.DataFrame([E.harness(run, label)['summary'] for label, run in a2_runs.items()]).set_index('variant')
a2[MODEL + SHOW]

,swing MSE vs count-only %,swing pred slope,take MSE vs count-only %,take pred slope,chase | zone_swing,zone_swing | chase,split-half r,YoY R2 (mean),Zone% |r|,next-season r (partial)
variant,,,,,,,,,,
v1 features,1.63,1.032,67.33,0.997,-0.920,0.700,0.819,0.596,0.299,0.085
+batter frame,1.76,1.049,69.84,0.998,-0.924,0.719,0.821,0.596,0.277,0.104
+handedness,1.76,1.036,70.06,0.996,-0.916,0.703,0.820,0.595,0.274,0.096
+pitch characteristics,1.91,1.037,70.21,0.998,-0.912,0.703,0.818,0.588,0.266,0.099


In [7]:
# The selection rule, rung by rung: each step against the one before, on the same pitches.
labels = list(a2_runs)
pd.concat({f'{b}  vs  {a}': pooled(a2_runs[b], a2_runs[a], ('new', 'old'))
           for a, b in zip(labels[:-1], labels[1:])}, names=['step']).round(3)

pitches  diff_%  diff_%_lo  diff_%_hi     verdict
step                                    action                                                   
+batter frame  vs  v1 features          take    1108923  -7.687     -8.113     -7.151  new better
                                        swing   1013946  -0.123     -0.140     -0.107  new better
+handedness  vs  +batter frame          take    1108923  -0.712     -0.830     -0.594  new better
                                        swing   1013946  -0.003     -0.010      0.004         tie
+pitch characteristics  vs  +handedness take    1108861  -0.500     -0.695     -0.292  new better
                                        swing   1013874  -0.153     -0.174     -0.131  new better

**Every rung makes a better model.** On the same held-out pitches, and reading
the rule as "better on at least one action, worse on neither":

- the **batter frame** cuts take MSE by 7.7% and swing MSE by 0.12%;
- **handedness** cuts take MSE by 0.7%, with swings a tie;
- **pitch characteristics** cut take MSE by 0.5% and swing MSE by 0.15%.

Every interval excludes zero except handedness on swings. The swing gains look
small only because most of a swing's squared error is outcome luck, which the
paired comparison cancels — the intervals are far from zero.

This corrects an earlier reading. The location-binned check suggested pitch
characteristics added nothing to the swing model; it could not see them,
because it averages over everything within a location cell. Paired accuracy
can, and they help both models.

All four models are close to calibrated (swing slopes 1.03–1.05, take 1.00).
The player checks barely move, except Zone% contamination, which falls from
0.299 to 0.266, and next-season validity, which rises from 0.085 to 0.099.

### And the events are predictable even though the run value is not

This is the case for decomposing a swing into whiff / foul / in-play rather
than regressing run value directly. Whether a batted ball falls in is close to
luck; whether the hitter *misses* is not. Same generic folds and learner.

In [8]:
pd.concat({'location + count': E.whiff_auc(df, V1_FEATURES),
           'full pitch frame': E.whiff_auc(df, V3_FEATURES)}, names=['features']).round(4)

swings     auc  logloss  logloss_base_rate  improvement_%
features         season                                                           
location + count 2023    339208  0.7322   0.4873             0.5709        14.6473
                 2024    337782  0.7330   0.4833             0.5659        14.5925
                 2025    336956  0.7353   0.4806             0.5650        14.9347
full pitch frame 2023    339208  0.7655   0.4699             0.5709        17.6887
                 2024    337782  0.7682   0.4646             0.5659        17.9101
                 2025    336884  0.7700   0.4618             0.5650        18.2693

**Whiff probability predicts at AUC 0.77 held out**, and velocity and movement
add clearly over location and count in every season (AUC 0.73 → 0.77; log loss
14.6–14.9% → 17.7–18.3% better than the base rate). The intermediate events
carry real signal — the premise of the event decomposition tested in v4.

## A3 — what is the take model actually learning?

A take can only end three ways, and the target is the league run value of
`(outcome, count)`. Given the count, the only thing location can tell the model
is the probability of a called strike. If that is right, then `take_pred` should
be an exact affine function of `P(called strike)` within each count, with slope
`RE(called_strike, c) − RE(ball, c)`.

Checked on fold 3: models trained on 2021–24, scored on held-out 2025.

In [9]:
TR3, VA3 = E.GENERIC_FOLDS[-1]
rv3 = D.run_value_table(df, TR3)
is_take = lambda d: ~d.swing & d.outcome.isin(['ball','called_strike'])
tr = df[df.season.isin(TR3) & is_take(df)]
cs_model = train_early_stopped(tr, V1_FEATURES, (tr.outcome == 'called_strike').astype(int), E.BINARY)

va = run_a1.held[(run_a1.held.season == VA3) & is_take(run_a1.held)].copy()
va['p_cs'] = cs_model.predict(_dmatrix(va, V1_FEATURES))

rows = []
for c, g in va.groupby('count', observed=True):
    if len(g) < 2000: continue
    r = np.corrcoef(g.q_take, g.p_cs)[0, 1]
    rows.append({'count': c, 'n': len(g), 'R2': r**2,
                 'fitted slope': np.polyfit(g.p_cs, g.q_take, 1)[0],
                 'RE(CS)-RE(ball)': rv3.get(('called_strike', c)) - rv3.get(('ball', c))})
a3 = pd.DataFrame(rows).set_index('count')
print(f'median R2 = {a3.R2.median():.4f}   (threshold 0.95)')
a3.round(4)

median R2 = 0.9915   (threshold 0.95)


,n,R2,fitted slope,RE(CS)-RE(ball)
count,,,,
0-0,122883,0.9753,-0.0804,-0.0781
0-1,46317,0.9402,-0.0885,-0.0849
0-2,23586,0.9527,-0.1960,-0.1892
1-0,39077,0.9918,-0.1116,-0.1106
1-1,31957,0.9772,-0.1158,-0.1136
1-2,29434,0.9709,-0.2351,-0.2299
2-0,13594,0.9967,-0.1730,-0.1731
2-1,15068,0.9943,-0.1804,-0.1792
2-2,20944,0.9913,-0.3267,-0.3259


**Confirmed on a held-out season, under the shared learner.** Median R² is 0.991
(lowest 0.940, on 0-1) and the fitted slopes match the run-value gaps to within
a few thousandths — on 3-2, −0.612 against an expected −0.614.

The take model is a called-strike probability wearing a regression's clothes:
it learns `P(CS | location)` and multiplies it by run values it reads off the
count. So a hitter-specific contact feature should not help it. A4 tests that
by accuracy.

## A4 — personalization

The binary hot-zone flag is replaced by a continuous surface: the hitter's
expected exit velocity at the pitch location, kernel-smoothed and shrunk toward
the league by effective sample size. Measured year over year the surface
reproduces itself at r ≈ 0.66 against ≈ 0.30 for raw bins (EDA §6).

- **Fixed two-season prior window** for both the hull and the surface.
- **Swing model only**, since A3 shows the take model is a called-strike
  probability — and the surface is also run in both models, to check.
- **Personalized folds for every row.**

In [10]:
nitro, coverage = F.season_in_nitro(df, window=F.PRIOR_WINDOW)
hot = F.season_hot_zone(df)
key = ['game_pk', 'at_bat_number', 'pitch_number']
pdf = hot.merge(nitro[key + ['in_nitro']], on=key, how='left', validate='one_to_one')
assert pdf.in_nitro.notna().all()

print('feature stability across seasons (the drift check):')
pd.concat([pdf.groupby('season').in_nitro.mean().rename('in_nitro rate'),
           pdf.groupby('season').hot_zone.mean().rename('hot_zone mean'),
           coverage.coverage.rename('hull coverage')], axis=1).round(3)

feature stability across seasons (the drift check):


,in_nitro rate,hot_zone mean,hull coverage
season,,,
2022,0.145,88.772,0.793
2023,0.203,88.640,0.824
2024,0.207,88.684,0.866
2025,0.211,88.811,0.842


In [11]:
F0 = ['plate_x_bat','plate_z_norm','count']
A4 = {
  'no personalization':               (F0, True),
  '+binary hull (swing only)':        (F0 + ['in_nitro'], True),
  '+continuous surface (swing only)': (F0 + ['hot_zone'], True),
  '+continuous surface (both models)':(F0 + ['hot_zone'], False),
}
a4_runs = {label: E.run_folds(pdf, feats, PF, fit=fit_direct, swing_only=so)
           for label, (feats, so) in A4.items()}
a4 = pd.DataFrame([E.harness(run, label)['summary'] for label, run in a4_runs.items()]).set_index('variant')
a4[MODEL + SHOW]

,swing MSE vs count-only %,swing pred slope,take MSE vs count-only %,take pred slope,chase | zone_swing,zone_swing | chase,split-half r,YoY R2 (mean),Zone% |r|,next-season r (partial)
variant,,,,,,,,,,
no personalization,1.72,1.089,69.66,1.007,-0.925,0.720,0.822,0.592,0.276,0.103
+binary hull (swing only),1.74,1.063,69.66,1.007,-0.919,0.707,0.823,0.590,0.263,0.105
+continuous surface (swing only),1.78,1.048,69.66,1.007,-0.888,0.657,0.846,0.635,0.172,0.165
+continuous surface (both models),1.78,1.048,69.55,1.000,-0.883,0.652,0.848,0.635,0.163,0.167


In [12]:
pd.concat({
    'hull vs none':                  pooled(a4_runs['+binary hull (swing only)'], a4_runs['no personalization'], ('new', 'old')),
    'surface vs none':               pooled(a4_runs['+continuous surface (swing only)'], a4_runs['no personalization'], ('new', 'old')),
    'surface vs hull':               pooled(a4_runs['+continuous surface (swing only)'], a4_runs['+binary hull (swing only)'], ('new', 'old')),
    'surface both vs swing only':    pooled(a4_runs['+continuous surface (both models)'], a4_runs['+continuous surface (swing only)'], ('new', 'old')),
}, names=['comparison']).round(3)

pitches  diff_%  diff_%_lo  diff_%_hi     verdict
comparison                 action                                                   
hull vs none               take    1108923   0.000      0.000      0.000         tie
                           swing   1013943  -0.018     -0.032     -0.007  new better
surface vs none            take    1108923   0.000      0.000      0.000         tie
                           swing   1013943  -0.059     -0.087     -0.035  new better
surface vs hull            take    1108923   0.000      0.000      0.000         tie
                           swing   1013943  -0.041     -0.066     -0.017  new better
surface both vs swing only take    1108923   0.354      0.138      0.607  old better
                           swing   1013943   0.000      0.000      0.000         tie

In [13]:
# Guardrails: the surface against no personalization, same hitters, hitter-resampled intervals.
E.metric_differences(a4_runs['+continuous surface (swing only)'], a4_runs['no personalization'],
                     labels=('surface', 'none')).round(3)

,surface,none,diff,diff_lo,diff_hi
check,,,,,
chase | zone_swing,-0.888,-0.925,0.037,0.028,0.048
zone_swing | chase,0.657,0.720,-0.063,-0.084,-0.043
YoY R2 (mean),0.635,0.592,0.042,0.021,0.061
Zone% |r|,0.172,0.276,-0.104,-0.123,-0.082
next-season r (partial),0.165,0.103,0.063,0.038,0.086


**The fixed window removes the drift.** The hull flag fires on 20.3–21.1% of
pitches from 2023 on, where the expanding window in v2 climbed to 25.1%.

By paired accuracy on the same pitches:

- **the binary hull** improves the swing model by 0.018% of MSE — detectable,
  and negligible;
- **the continuous surface** improves it by 0.059% (interval 0.035–0.087%),
  three times the hull, and beats the hull head to head;
- **putting the surface in the take model as well makes the take model worse**
  (by 0.35%, interval 0.14–0.61%). The rule, not just the argument from A3, now
  says hitter features belong on the swing side only.

**The guardrails move in both directions.** Against no personalization, the
surface raises YoY R² (+0.042) and next-season validity (+0.063) and cuts Zone%
contamination (−0.104), all with intervals excluding zero — but construct
validity slips (−0.925 / +0.720 → −0.888 / +0.657), also outside its interval.
A better estimate of what a swing is worth to this hitter partly moves the
score toward "is he a good hitter". Guardrails flag that; they do not overrule
the more accurate model.

### Does the hot zone actually differ between hitters?

"Redundant with location" would mean every hitter's hot zone is the same
region — in which case the model already knows it from `plate_x_bat` and
`plate_z_norm`. Check the premise directly.

In [14]:
inz = D.in_rulebook_zone(pdf, ball_edge=True)
z = pdf[inz].assign(
    cx=pd.cut(pdf[inz].plate_x_bat, np.linspace(-0.85, 0.85, 5), labels=False),
    cz=pd.cut(pdf[inz].plate_z_norm, np.linspace(0, 1, 5), labels=False)).dropna(subset=['cx','cz'])
cellwise = z.groupby(['season','cx','cz','batter'], observed=True).hot_zone.mean().rename('v').reset_index()
spread = cellwise.groupby(['season','cx','cz'], observed=True).v.agg(['mean','std','size']).query('size >= 100')

print(f'between-hitter SD within one zone cell : {spread["std"].median():.2f} mph')
print(f'league mean across cells               : {spread["mean"].min():.1f} - {spread["mean"].max():.1f} mph')
print(f'  -> personalization is {spread["std"].median()/(spread["mean"].max()-spread["mean"].min()):.1f}x the size of the location effect\n')

best = cellwise.groupby(['season','batter'], observed=True).apply(
    lambda g: g.loc[g.v.idxmax(), ['cx','cz']], include_groups=False)
print("where each hitter's best cell falls (share of hitter-seasons):")
(best.groupby(['cx','cz']).size() / len(best)).unstack().round(3)

between-hitter SD within one zone cell : 1.72 mph
league mean across cells               : 87.3 - 90.2 mph
  -> personalization is 0.6x the size of the location effect



where each hitter's best cell falls (share of hitter-seasons):


cz,0.0,1.0,2.0,3.0
cx,,,,
0.0,0.194,0.027,0.036,0.035
1.0,0.045,0.146,0.211,0.075
2.0,0.063,0.097,0.042,0.008
3.0,0.015,0.003,0.002,0.002


**The premise fails.** Hitters differ at the same location by 1.72 mph — about
0.6× the entire league-wide location effect (87.3–90.2 mph) — and they differ
in *where* their best region is: the most common best cell holds only 21% of
hitter-seasons. Hitters cannot cover the whole zone, and they do not cover the
same part of it.

In [15]:
# How much does the feature move the counterfactual, and how often does it flip
# the recommended action? Fold 3 models, held-out 2025.
va = pdf[pdf.season == 2025]
d_no  = predict_both(va, a4_runs['no personalization'].models[2025])
d_hot = predict_both(va, a4_runs['+continuous surface (swing only)'].models[2025])

dq = d_hot.q_swing - d_no.q_swing
flip = np.sign(d_no.edge) != np.sign(d_hot.edge)
print(f'change in Q_swing from the feature : SD {dq.std():.4f} runs  (Q_swing itself: SD {d_no.q_swing.std():.4f})')
print(f'change in Q_take                   : max |.| {(d_hot.q_take - d_no.q_take).abs().max():.4f} runs  (take model has no hot_zone)')
print(f'pitches where the RECOMMENDED ACTION flips: {flip.mean():.2%}')

change in Q_swing from the feature : SD 0.0125 runs  (Q_swing itself: SD 0.0468)
change in Q_take                   : max |.| 0.0000 runs  (take model has no hot_zone)
pitches where the RECOMMENDED ACTION flips: 2.00%


**A large effect that rarely flips the decision.** The surface moves `Q_swing`
with a standard deviation of 0.0125 runs, about a quarter of `Q_swing`'s own
(0.047), but flips the recommended action on only 2.0% of pitches, and `Q_take`
does not move at all. A sign-based score registers the feature on those 2% and
nowhere else.

In [16]:
# The feature against every metric, models held fixed.
rows = []
for feat_label, run_label in [('without hot_zone', 'no personalization'),
                              ('with hot_zone',    '+continuous surface (swing only)')]:
    for score in DEC.SCORES:
        rows.append({'metric': score, 'features': feat_label,
                     **E.metric_checks(a4_runs[run_label].held, score)['summary']})
inter = pd.DataFrame(rows).set_index(['metric','features'])
inter

,,chase | zone_swing,zone_swing | chase,split-half r,YoY R2 (mean),Zone% |r|,next-season r (partial)
metric,features,,,,,,
chosen_value,without hot_zone,-0.890,0.678,0.759,0.529,0.113,0.139
signed_edge,without hot_zone,-0.925,0.720,0.822,0.592,0.276,0.103
regret,without hot_zone,-0.856,0.581,0.782,0.556,0.494,0.051
close_weighted,without hot_zone,-0.740,0.813,0.620,0.296,0.053,0.117
correct_decision,without hot_zone,-0.964,0.926,0.833,0.585,0.239,0.135
chosen_value,with hot_zone,-0.601,0.404,0.913,0.713,0.178,0.292
signed_edge,with hot_zone,-0.888,0.657,0.846,0.635,0.172,0.165
regret,with hot_zone,-0.785,0.446,0.806,0.566,0.557,-0.021
close_weighted,with hot_zone,-0.641,0.729,0.622,0.253,0.125,0.062


**Personalization interacts with the metric** — with the model held fixed, so
this is about the score, not accuracy.

- Under `chosen_value` construct validity collapses (−0.890 / +0.678 →
  −0.601 / +0.404) while reliability and next-season validity soar
  (0.139 → 0.292): the score drifts toward hitting ability.
- Under `signed_edge` reliability rises, Zone% contamination falls by 0.104,
  next-season validity rises by 0.062, and construct validity gives back
  0.04–0.06.
- Under `correct_decision` construct validity and reliability hardly move,
  and next-season validity rises 0.135 → 0.155.

`correct_decision` keeps by far the best construct validity (−0.957 / +0.911
with the surface). `signed_edge` with the surface is more reliable (YoY 0.635
against 0.580) and less contaminated (Zone% 0.172 against 0.220). The same
trade as in A1.

## The ladder at a fixed metric

v1 and v2 score `chosen_value` because that is their design, while this
notebook changed the metric as well as the features. So the whole feature
ladder is run with the metric held fixed, on the personalized folds, under both
candidate metrics. Each row is a complete feature set, not an increment: the
hull and the surface estimate the same thing, so they are never combined.
Hitter features go to the swing model only, except in the last row.

In [17]:
LADDER = {
  'v1  location+count':                  (V1_FEATURES, True),
  'v2  +nitro hull':                     (V1_FEATURES + ['in_nitro'], True),
  'v3  full pitch frame':                (V3_FEATURES, True),
  'v3  +hot-zone surface':               (V3_FEATURES + ['hot_zone'], True),
  'v3  +hot-zone surface (both models)': (V3_FEATURES + ['hot_zone'], False),
}
ladder_runs = {label: E.run_folds(pdf, feats, PF, fit=fit_direct, swing_only=so, in_sample=True)
               for label, (feats, so) in LADDER.items()}
rows = []
for score in ['chosen_value', 'signed_edge']:
    for label, run in ladder_runs.items():
        rows.append({'metric': score, 'features': label, **E.harness(run, label, score)['summary']})
ladder = pd.DataFrame(rows).drop(columns='variant').set_index(['metric','features'])
ladder.loc['signed_edge', MODEL]

,swing MSE vs count-only %,swing pred slope,take MSE vs count-only %,take pred slope
features,,,,
v1 location+count,1.61,1.067,67.23,1.002
v2 +nitro hull,1.63,1.059,67.23,1.002
v3 full pitch frame,1.87,1.053,70.08,0.997
v3 +hot-zone surface,1.93,1.025,70.08,0.997
v3 +hot-zone surface (both models),1.93,1.025,69.98,0.996


In [18]:
ladder[SHOW]

chase | zone_swing  zone_swing | chase  split-half r  YoY R2 (mean)  Zone% |r|  next-season r (partial)
metric       features                                                                                                                                    
chosen_value v1  location+count                               -0.890               0.678         0.763          0.538      0.116                    0.129
             v2  +nitro hull                                  -0.812               0.557         0.793          0.518      0.058                    0.132
             v3  full pitch frame                             -0.875               0.654         0.743          0.478      0.139                    0.109
             v3  +hot-zone surface                            -0.602               0.403         0.911          0.696      0.166                    0.286
             v3  +hot-zone surface (both models)              -0.601               0.403         0.911          0.697      0.167                    0.284
signed_edge  v1  location+count                               -0.921               0.703         0.819          0.594      0.298                    0.084
             v2  +nitro hull                                  -0.915               0.692         0.820          0.589      0.282                    0.087
             v3  full pitch frame                             -0.916               0.711         0.815          0.584      0.269                    0.104
             v3  +hot-zone surface                            -0.885               0.662         0.844          0.629      0.169                    0.167
             v3  +hot-zone surface (both models)              -0.882               0.657         0.844          0.630      0.163                    0.168

In [19]:
# The selection rule along the ladder.
STEPS = [('v2  +nitro hull', 'v1  location+count'),
         ('v3  full pitch frame', 'v1  location+count'),
         ('v3  +hot-zone surface', 'v3  full pitch frame'),
         ('v3  +hot-zone surface (both models)', 'v3  +hot-zone surface')]
pd.concat({f'{a}  vs  {b}': pooled(ladder_runs[a], ladder_runs[b], ('new', 'old')) for a, b in STEPS},
          names=['step']).round(3)

pitches  diff_%  diff_%_lo  diff_%_hi     verdict
step                                               action                                                   
v2  +nitro hull  vs  v1  location+count            take    1108923   0.000      0.000      0.000         tie
                                                   swing   1013943  -0.023     -0.037     -0.012  new better
v3  full pitch frame  vs  v1  location+count       take    1108861  -8.689     -9.177     -8.097  new better
                                                   swing   1013871  -0.264     -0.294     -0.237  new better
v3  +hot-zone surface  vs  v3  full pitch frame    take    1108861   0.000      0.000      0.000         tie
                                                   swing   1013871  -0.062     -0.090     -0.039  new better
v3  +hot-zone surface (both models)  vs  v3  +h... take    1108861   0.346      0.106      0.591  old better
                                                   swing   1013871   0.000      0.000      0.000         tie

In [20]:
# Guardrails along the ladder under signed_edge, hitter-resampled intervals.
pd.concat({f'{a}  vs  {b}': E.metric_differences(ladder_runs[a], ladder_runs[b], labels=('new', 'old'))
           for a, b in STEPS}, names=['step']).round(3)

new    old   diff  diff_lo  diff_hi
step                                               check                                                         
v2  +nitro hull  vs  v1  location+count            chase | zone_swing      -0.915 -0.921  0.006    0.003    0.008
                                                   zone_swing | chase       0.692  0.703 -0.011   -0.019   -0.003
                                                   YoY R2 (mean)            0.589  0.594 -0.005   -0.012    0.002
                                                   Zone% |r|                0.282  0.298 -0.016   -0.022   -0.009
                                                   next-season r (partial)  0.087  0.084  0.004   -0.005    0.013
v3  full pitch frame  vs  v1  location+count       chase | zone_swing      -0.916 -0.921  0.005   -0.001    0.010
                                                   zone_swing | chase       0.711  0.703  0.008   -0.009    0.025
                                                   YoY R2 (mean)            0.584  0.594 -0.010   -0.029    0.005
                                                   Zone% |r|                0.269  0.298 -0.028   -0.044   -0.016
                                                   next-season r (partial)  0.104  0.084  0.020    0.005    0.037
v3  +hot-zone surface  vs  v3  full pitch frame    chase | zone_swing      -0.885 -0.916  0.031    0.021    0.041
                                                   zone_swing | chase       0.662  0.711 -0.049   -0.073   -0.027
                                                   YoY R2 (mean)            0.629  0.584  0.045    0.026    0.064
                                                   Zone% |r|                0.169  0.269 -0.100   -0.118   -0.077
                                                   next-season r (partial)  0.167  0.104  0.063    0.039    0.085
v3  +hot-zone surface (both models)  vs  v3  +h... chase | zone_swing      -0.882 -0.885  0.003    0.002    0.004
                                                   zone_swing | chase       0.657  0.662 -0.005   -0.007   -0.003
                                                   YoY R2 (mean)            0.630  0.629  0.000   -0.003    0.003
                                                   Zone% |r|                0.163  0.169 -0.007   -0.009   -0.005
                                                   next-season r (partial)  0.168  0.167  0.001   -0.001    0.003

**The selection rule, down the ladder:**

- **v2's hull** makes the swing model better by 0.023% — detectable, negligible.
- **v3's pitch frame** beats v1's location and count on both models (take
  −8.7%, swing −0.26%).
- **The hot-zone surface** improves the pitch-frame model's swing side by
  0.062% (0.039–0.090%).
- **Giving the surface to the take model too** makes it worse (+0.35%).

So **v3 is the full pitch frame with the surface in the swing model only.** Its
swing model is also the best calibrated on the ladder (slope 1.025, against
1.05–1.07 for the rungs below it).

**Guardrails along the way.** The pitch frame improves next-season validity
(+0.020) and Zone% (−0.028) with construct validity unchanged. The surface
supplies the larger gains (YoY +0.045, Zone% −0.100, next-season +0.063) and
the construct cost (−0.916 / +0.711 → −0.885 / +0.662). Under `chosen_value`
the same surface collapses construct validity, which is why the metric matters.

## v3

The rule selected full pitch frame + hot-zone surface (swing only). Step 2:
check its calibration and, where a slope interval excludes 1, recalibrate with
a linear map fitted on each fold's early-stopping games.

In [21]:
SELECTED = 'v3  +hot-zone surface'        # chosen by the paired-accuracy table above
v3_run = ladder_runs[SELECTED]
cal = E.calibration_table(v3_run)
needs = bool(((cal.slope_lo > 1) | (cal.slope_hi < 1)).any())
print(f'selected: {SELECTED}')
print(f'recalibration needed (some slope interval excludes 1): {needs}')
cal.round(3)

selected: v3  +hot-zone surface
recalibration needed (some slope interval excludes 1): True


slope  slope_lo  slope_hi
action season                           
take   2023    1.000     0.996     1.003
       2024    0.998     0.994     1.002
       2025    0.994     0.991     0.998
swing  2023    1.054     1.026     1.077
       2024    1.009     0.988     1.031
       2025    1.012     0.990     1.034

In [22]:
# Step 2 of the rule: recalibrate on each fold's early-stopping games (never the held-out season).
feats, so = LADDER[SELECTED]
v3_final = (E.run_folds(pdf, feats, PF, fit=fit_direct, swing_only=so, recalibrate=True, in_sample=True)
            if needs else v3_run)
print('calibration maps (a, b) per fold:')
print(pd.DataFrame({s: {f'{k} {x}': v for k, ab in (m.calibration or {}).items() for x, v in zip('ab', ab)}
                    for s, m in v3_final.models.items()}).T.round(4))
pd.concat({'as fitted': cal, 'recalibrated': E.calibration_table(v3_final)}, names=['model']).round(3)

calibration maps (a, b) per fold:
      take a  take b  swing a  swing b
2023  0.0002  0.9999   0.0012    0.972
2024 -0.0002  1.0023  -0.0018    0.970
2025  0.0001  0.9963  -0.0021    1.010


slope  slope_lo  slope_hi
model        action season                           
as fitted    take   2023    1.000     0.996     1.003
                    2024    0.998     0.994     1.002
                    2025    0.994     0.991     0.998
             swing  2023    1.054     1.026     1.077
                    2024    1.009     0.988     1.031
                    2025    1.012     0.990     1.034
recalibrated take   2023    1.000     0.996     1.003
                    2024    0.996     0.992     0.999
                    2025    0.998     0.994     1.001
             swing  2023    1.084     1.056     1.108
                    2024    1.040     1.019     1.063
                    2025    1.002     0.980     1.024

In [23]:
# What recalibration did to accuracy and to the score.
display(pooled(v3_final, v3_run, ('recalibrated', 'as fitted')).round(3))
v3 = E.harness(v3_final, 'v3')
before = E.harness(v3_run, 'v3 as fitted')
ins = E.in_sample_checks(v3_final, 'v3')
pd.DataFrame([before['summary'], v3['summary'], ins['summary']]).set_index('variant')[MODEL + SHOW]

,pitches,diff_%,diff_%_lo,diff_%_hi,verdict
action,,,,,
take,1108861,-0.012,-0.015,-0.008,recalibrated better
swing,1013871,-0.001,-0.003,0.002,tie


,swing MSE vs count-only %,swing pred slope,take MSE vs count-only %,take pred slope,chase | zone_swing,zone_swing | chase,split-half r,YoY R2 (mean),Zone% |r|,next-season r (partial)
variant,,,,,,,,,,
v3 as fitted,1.93,1.025,70.08,0.997,-0.885,0.662,0.844,0.629,0.169,0.167
v3,1.93,1.042,70.09,0.998,-0.887,0.660,0.844,0.630,0.172,0.166
v3 (in-sample),NaN,NaN,NaN,NaN,-0.898,0.658,0.843,0.645,0.215,0.166


In [24]:
# Does Q_swing hold where it is extrapolated? Swings at pitches the league usually takes.
p_swing = E.swing_propensity(pdf, V3_FEATURES, PF)
E.propensity_calibration(v3_final, p_swing).round(4)

pitches  mean pred  mean obs   slope  slope_lo  slope_hi
action league swing propensity                                                          
swing  (-0.001, 0.1]              13481    -0.0591   -0.0666  1.0010    0.9221    1.0794
       (0.1, 0.3]                 70228    -0.0507   -0.0515  1.1156    1.0781    1.1529
       (0.3, 0.7]                356086    -0.0300   -0.0313  1.0441    1.0182    1.0648
       (0.7, 0.9]                322603    -0.0210   -0.0200  1.0119    0.9799    1.0399
       (0.9, 1.0]                251473    -0.0029   -0.0005  1.0262    0.9883    1.0629
take   (-0.001, 0.1]             363234     0.0530    0.0536  0.9926    0.9883    0.9967
       (0.1, 0.3]                290224     0.0282    0.0292  0.9967    0.9932    1.0003
       (0.3, 0.7]                361843     0.0002    0.0010  1.0007    0.9975    1.0039
       (0.7, 0.9]                 81067    -0.0394   -0.0381  1.0132    1.0034    1.0213
       (0.9, 1.0]                 12493    -0.1353   -0.1339  1.0368    1.0212    1.0506

In [25]:
try:
    from pybaseball import playerid_reverse_lookup
    s25 = v3['scores'].query('season == 2025').copy()
    nm = playerid_reverse_lookup(s25.batter.tolist())
    nm['name'] = (nm.name_first + ' ' + nm.name_last).str.title()
    s25 = s25.merge(nm[['key_mlbam','name']], left_on='batter', right_on='key_mlbam')
    cols = ['name','pitches','decision_value']
    print('TOP 10, 2025');    display(s25.nlargest(10, 'decision_value')[cols].round(1))
    print('BOTTOM 10, 2025'); display(s25.nsmallest(10, 'decision_value')[cols].round(1))
except Exception as exc:
    print(f'name lookup unavailable ({exc})')

Gathering player lookup table. This may take a moment.


TOP 10, 2025


,name,pitches,decision_value
139,Ronald Acuña,1741,130.3
142,Leo Rivas,520,129.4
116,Gleyber Torres,2652,125.4
31,Max Muncy,1603,125.1
280,Davis Schneider,1012,124.9
97,Matt Thaiss,768,124.0
154,Kyle Tucker,2275,122.7
39,Aaron Judge,2631,122.6
180,Juan Soto,2968,121.4
301,Max Schuemann,838,121.1


BOTTOM 10, 2025


,name,pitches,decision_value
47,Javier Báez,1543,65.6
83,Edmundo Sosa,942,70.8
296,Jhonkensy Noel,563,71.3
326,Angel Martínez,1793,73.2
120,Yohel Pozo,591,74.6
268,Lenyn Sosa,2056,75.2
286,Ronny Mauricio,692,75.2
0,Martín Maldonado,569,76.4
256,Michael Harris,2173,78.6
81,Reese Mcguire,503,79.1


## Contamination — how much of the score is opportunity?

`signed_edge` pays more for an obvious take than for a close call (A1), so a
hitter thrown more junk can score higher for the same judgement. Raw Zone% |r|
is a screen for that, and a confounded one: better hitters are thrown fewer
strikes, so a metric that identifies good hitters picks up a *negative* Zone%
pull that can mask a positive bias.

So Zone% is also reported holding something fixed. Every control is a screen,
not proof:

- **correct-decision rate** is computed from the same model's Δ, and for
  `correct_decision` itself it is the metric, so that cell is meaningless;
- **chase and zone-swing rates** are model-free, but they vary with how hard the
  pitches a hitter sees are, so they do not hold judgement fixed either;
- **production** (ΔRE per PA) removes the "better hitters see fewer strikes"
  pull, which is what makes the raw test lenient.

The only test with ground truth is a known-policy simulation (Track B3).

In [26]:
key = ['season', 'batter']

def hitter_table(held):
    held = held.assign(inz=D.in_rulebook_zone(held, ball_edge=True), right=held.correct_decision > 0)
    g = held.groupby(key, observed=True)
    t = pd.DataFrame({'zone_pct': g.inz.mean(), 'cd_rate': g.right.mean(),
                      'chase': held[~held.inz].groupby(key, observed=True).swing.mean(),
                      'zone_swing': held[held.inz].groupby(key, observed=True).swing.mean()}).reset_index()
    return t.merge(E._production(held), on=key)

def partial(y, x, controls):
    Z = np.column_stack([np.ones(len(y))] + [np.asarray(c, float) for c in controls])
    res = lambda v: v - Z @ np.linalg.lstsq(Z, np.asarray(v, float), rcond=None)[0]
    return float(np.corrcoef(res(y), res(x))[0, 1])

CONTROLS = {'raw': [], '| correct-decision rate': ['cd_rate'],
            '| chase, zone-swing': ['chase', 'zone_swing'], '| production': ['re_per_pa']}

def contamination(scores, hit):
    """Zone% correlation per held-out season, averaged, under each control."""
    m = scores.merge(hit, on=key).dropna(subset=['decision_value', 'zone_pct', *hit.columns[2:]])
    return {f'Zone% r {label}': np.mean([partial(s.decision_value, s.zone_pct, [s[c] for c in ctrl])
                                         for _, s in m.groupby('season')])
            for label, ctrl in CONTROLS.items()}

hv3 = v3_final.held
hit_v3, hit_v1 = hitter_table(hv3), hitter_table(run_a1.held)
rows = []
for model, held, hit in [('v1 features', run_a1.held, hit_v1), ('v3', hv3, hit_v3)]:
    for score in ['chosen_value', 'signed_edge', 'regret', 'correct_decision']:
        rows.append({'model': model, 'score': score, **contamination(E.player_metric(held, score), hit)})
contam = pd.DataFrame(rows).set_index(['model', 'score'])
print(f"corr(Zone%, production) across hitter-seasons: {np.corrcoef(hit_v3.zone_pct, hit_v3.re_per_pa)[0, 1]:+.3f}")
contam.round(3)

corr(Zone%, production) across hitter-seasons: -0.225


Zone% r raw  Zone% r | correct-decision rate  Zone% r | chase, zone-swing  Zone% r | production
model       score                                                                                                            
v1 features chosen_value            0.124                           -0.201                       -0.359                 0.248
            signed_edge             0.299                            0.172                       -0.006                 0.396
            regret                  0.504                            0.561                        0.470                 0.569
            correct_decision        0.250                            0.060                        0.160                 0.337
v3          chosen_value           -0.165                           -0.421                       -0.458                -0.037
            signed_edge             0.172                           -0.072                       -0.252                 0.301
            regret                  0.555                            0.599                        0.533                 0.591
            correct_decision        0.230                            0.036                        0.039                 0.328

In [27]:
# Rescaling cannot help: z-score, an OPS+-style ratio and a percentile rank are
# monotone transforms of the same per-hitter mean.
sc = E.player_metric(hv3, 'signed_edge')
sc['ratio'] = sc.raw / sc.groupby('season').raw.transform('mean') * 100
sc['percentile'] = sc.groupby('season').raw.rank(pct=True) * 100
print(sc[['decision_value', 'ratio', 'percentile']].corr(method='spearman').round(4))
pd.DataFrame([{'scaling': col, **contamination(sc.assign(decision_value=sc[col]), hit_v3)}
              for col in ['decision_value', 'ratio', 'percentile']]).set_index('scaling').round(3)

                decision_value   ratio  percentile
decision_value          1.0000  0.9999      0.9997
ratio                   0.9999  1.0000      0.9997
percentile              0.9997  0.9997      1.0000


,Zone% r raw,Zone% r | correct-decision rate,"Zone% r | chase, zone-swing",Zone% r | production
scaling,,,,
decision_value,0.172,-0.072,-0.252,0.301
ratio,0.172,-0.072,-0.252,0.301
percentile,0.165,-0.066,-0.240,0.287


In [28]:
# Departure weighting: (swung - P(league swings | pitch)) x edge, so a pitch
# everyone handles the same way contributes ~0 to everyone. P(swing) is the
# league propensity fitted per fold on that fold's training seasons.
hv3 = hv3.join(p_swing, on=E.PITCH_KEY)
hv3['departure'] = (hv3.swing.astype(float) - hv3.p_swing) * hv3.edge
hv3['region'] = region(hv3)
print(hv3.groupby('region', observed=True)[['signed_edge', 'departure']].mean().round(4), '\n')
pd.DataFrame({s: {**E.metric_checks(hv3, s)['summary'],
                  **contamination(E.player_metric(hv3, s), hit_v3)}
              for s in ['signed_edge', 'departure']}).round(3)

        signed_edge  departure
region                        
chase        0.0814     0.0010
shadow       0.0089     0.0006
zone         0.0498    -0.0001 



,signed_edge,departure
chase | zone_swing,-0.887,-0.899
zone_swing | chase,0.660,0.658
split-half r,0.844,0.864
YoY R2 (mean),0.630,0.652
Zone% |r|,0.172,0.327
next-season r (partial),0.166,0.100
Zone% r raw,0.172,0.327
Zone% r | correct-decision rate,-0.072,0.269
"Zone% r | chase, zone-swing",-0.252,0.093
Zone% r | production,0.301,0.424


In [29]:
# Opportunity standardization: a hitter's mean signed_edge within each stratum,
# reweighted to the league's mix of strata that season. Strata a hitter never
# saw take the league mean.
FAMILY = {**dict.fromkeys(['FF', 'SI', 'FC'], 'fastball'),
          **dict.fromkeys(['SL', 'ST', 'SV', 'CU', 'KC', 'CS'], 'breaking'),
          **dict.fromkeys(['CH', 'FS', 'FO', 'SC', 'KN', 'EP'], 'offspeed')}
hv3['family'] = hv3.pitch_type.astype(str).map(FAMILY).fillna('other')

def standardized(strata):
    cell = hv3.groupby(key + strata, observed=True).signed_edge.mean().rename('m').reset_index()
    league = hv3.groupby(['season'] + strata, observed=True).signed_edge.agg(lm='mean', n='size').reset_index()
    league['w'] = league.n / league.groupby('season').n.transform('sum')
    grid = hit_v3[key].merge(league, on='season').merge(cell, on=key + strata, how='left')
    grid['m'] = grid.m.fillna(grid.lm)
    s = (grid.w * grid.m).groupby([grid.season, grid.batter]).sum().rename('raw').reset_index()
    qualified = E.player_metric(hv3, 'signed_edge')[key]          # same >=500-pitch hitters
    s = s.merge(qualified, on=key)
    z = s.groupby('season').raw.transform(lambda v: (v - v.mean()) / v.std(ddof=0))
    return s.assign(decision_value=100 + 10 * z)

rows = [{'scoring': 'unadjusted', **contamination(E.player_metric(hv3, 'signed_edge'), hit_v3),
         **E.construct_validity(hv3, E.player_metric(hv3, 'signed_edge')).iloc[0][['chase | zone_swing', 'zone_swing | chase']]}]
for label, strata in [('region x count', ['region', 'count']),
                      ('region x count x pitch family', ['region', 'count', 'family'])]:
    s = standardized(strata)
    rows.append({'scoring': f'standardized: {label}', **contamination(s, hit_v3),
                 **E.construct_validity(hv3, s).iloc[0][['chase | zone_swing', 'zone_swing | chase']]})
pd.DataFrame(rows).set_index('scoring').T.round(3)

scoring,unadjusted,standardized: region x count,standardized: region x count x pitch family
Zone% r raw,0.172,0.224,0.204
Zone% r | correct-decision rate,-0.072,0.049,0.009
"Zone% r | chase, zone-swing",-0.252,-0.102,-0.122
Zone% r | production,0.301,0.350,0.333
chase | zone_swing,-0.887,-0.866,-0.859
zone_swing | chase,0.660,0.620,0.626


**The answer depends on which control you pick, so no control settles it.**
For v3's `signed_edge`, Zone% correlates +0.172 raw, −0.072 holding the
correct-decision rate fixed, −0.252 holding chase and zone-swing rates fixed,
and +0.301 holding production fixed.

**The raw test is lenient.** Better hitters are thrown fewer strikes
(corr(Zone%, production) = −0.225), so holding production fixed raises every
score's Zone% correlation.

**`correct_decision` is not cleaner on the final model.** On v1's features it
is (raw 0.250 against 0.299; 0.337 against 0.396 holding production fixed). On
v3 the surface closes the gap: 0.230 against 0.172 raw, 0.328 against 0.301
holding production fixed.

**Rescaling cannot fix it** — z-score, ratio and percentile correlate 0.9997
or more.

**Departure weighting fails**: the most reliable score tested (split-half
0.864, YoY 0.652), but raw Zone% rises to 0.327 and next-season validity falls
to 0.100.

**Opportunity standardization does not remove it.** Reweighting to the league
mix of region × count (± pitch family) raises raw Zone% (0.172 → 0.224 /
0.204) and the production-held figure (0.301 → 0.350 / 0.333) and costs
construct validity; only the behaviour-held figure moves toward zero. The mix
across these strata is not what drives it; mix at a finer grain is untested.

## Findings

All figures held out, under the shared early-stopping learner. Models chosen by
paired held-out accuracy; the player checks are guardrails.

**Every design change is a better model.** The batter frame, handedness and
pitch characteristics each improve held-out accuracy — together cutting take
MSE by 8.7% and swing MSE by 0.26% against v1's location and count, every
interval excluding zero. The hot-zone surface improves the swing model a
further 0.062%; the binary hull, a third as much. Putting the surface in the
take model makes that model worse, confirming by accuracy what A3 argued.

**v3 = full pitch frame + hot-zone surface, swing model only, `signed_edge`.**
Held out: construct −0.887 / +0.660, split-half 0.844, YoY R² 0.630, Zone% |r|
0.172, next-season partial r 0.166.

**Calibration: close, and recalibration did not help.** As fitted, the take
model's slopes are 0.99–1.00 and the swing model's 1.01–1.05; the intervals
exclude 1 for takes in 2025 and swings in 2023, so the rule recalibrated.
Recalibration made takes marginally better (MSE −0.012%) but swings worse
calibrated in 2023 and 2024 (1.054 → 1.084, 1.009 → 1.040), with swing MSE
unchanged. The map is fitted on held-back games from the *training* seasons,
while the remaining miscalibration is a shift between seasons, which such a map
cannot see. The v3 figures above are the recalibrated model's; the as-fitted
model's player checks differ by at most 0.003.

**`Q_swing` is slightly too generous where it is extrapolated.** On swings at
pitches the league swings at less than 10% of the time it predicts −0.059 runs
against −0.067 observed. Chases of the most obvious balls are therefore
penalised slightly less than they should be.

**The metric is a trade, not a win.** By the criterion fixed in advance
`correct_decision` has the better construct validity; `signed_edge` keeps the
run-value magnitude and is the score under which the more accurate,
personalized model also improves the guardrails. Its pitch-mix contamination is
real and unsolved.

**Face validity holds.** The 2025 leaderboard puts Acuña, Muncy, Tucker, Judge
and Soto near the top and Báez at the bottom; the extremes include hitters just
over the 500-pitch line, where the score is noisiest.

In [30]:
pd.DataFrame([v3['summary']]).set_index('variant')

,swing MSE vs count-only %,swing pred slope,take MSE vs count-only %,take pred slope,chase | zone_swing,zone_swing | chase,split-half r,YoY R2 (mean),Zone% |r|,next-season r (partial)
variant,,,,,,,,,,
v3,1.93,1.042,70.09,0.998,-0.887,0.66,0.844,0.63,0.172,0.166
